text/0091-ci-upload-tokens.md

```diff
@@ -0,0 +1,273 @@
+routing to the right URLs.  Today the documentation is very quiet about how to get this
+system to work on a single tenant installation and customers are often required to work
+with CS to get source maps working.
+
+# Technical Implementation
+
+The motivation is to add a new kind of token to Sentry which are fundamentally per-organization
+tokens, but with the ability to carry meta information that tools like sentry-cli can use to
+improve the user experience.  These org level tokens can be created by anyone in the org, they
+can be given additional restrictions, and they can carry meta information such as routing
+data.  For the purpose of this document they are called **structural tokens**.
+
+## Token Format
+
+The proposed token format is undecided so far.  The goals of the token align generally with
+both [Macaroons](http://macaroons.io/) and [Biscuit](https://www.biscuitsec.org).  Unfortunately
+the former standard has never seen much attention, and the latter is pretty new and not
+particularly proven.  Either system however permits adding additional restrictions to the
+token which make them a perfect choice for the use in our pipeline.  Biscuits however seem
+quite promising.  The idea of biscuits is that the token itself holds _facts_ which and
+can be further constraints via _checks_ and they are checked against a datalog inspired
+_policies_.
+
+One of the benefits of having the tokens carry this data is that the token alone has enough
+information available to route to a Sentry installation.  This means that `sentry-cli` or
+any other tool _just_ needs the token to even determine the host that the token should be
+sent against.  This benefit also applies to JWT or PASETO tokens which can be considered
+for this as well.  The RFC here thus proposes two potential options: A **Biscuit** token
+format and a **JWT** token format.
+
+A serialized token is added a custom prefix `sntrys_` (sentry structure) to make
+it possible to detect it by security scrapers.  Anyone handling such a token is
+required to check for the `sntrys_` prefix and disregard it before parsing it.  This
+can also be used by the client side to detect a structural token if the client is
+interested in extracting data from the token.
+
+## Token Facts
+
+We want to encode certain information into the tokens.  In Biscuit terms these are called
+_facts_.  The following facts exist:
+
+* `site`: references the target API URL that should be used.  A token will always have a
+  site in it and clients are not supposed to provide a fallback.  For instance this
+  would be `https://myorg.sentry.io/`.
+* `org`: a token is uniquely bound to an org, so the slug of that org is also always
+  contained.  Note that the slug is used rather than an org ID as the clients typically
+  need these slugs to create API requests.
+* `projects`: a token can be valid for more than one project.  For operations such as
```

Comment:
we dont need project for the use case we have - so do we really need that as part of this spec?

Reply:
Having the project as part of the token would make it possible that _just_ the token is all that is needed to upload debug files or source maps and automatically associate it with the right project without further configuration needed.  Otherwise it's yet again an extra piece of information that needs to be configured separately.